docs/rfcs/826-add-multipart-http-signer-support.md

```diff
@@ -0,0 +1,78 @@
+- Author: abyss-w <mad.hatter@foxmail.com>
+- Start Date: 2021-09-26
+- RFC PR: [beyondstorage/go-storage#826](https://github.com/beyondstorage/go-storage/pull/826)
+- Tracking Issue: [beyondstorage/go-storage#0](https://github.com/beyondstorage/go-storage/issues/0)
+
+# GSP-826: Add Multipart HTTP Signer Support
+
+Previous discussion:
+
+- [Multipart related operations ](https://forum.beyondstorage.io/t/topic/226)
+
+# Background
+
+In [GSP-729](https://github.com/beyondstorage/go-storage/blob/master/docs/rfcs/729-redesign-http-signer.md), we split out the `HTTPSigner` proposed in [GSP-706](https://github.com/beyondstorage/go-storage/blob/master/docs/rfcs/706-support-http-signer.md). We had the following problems in the implementation of GSP-706:
+
+- There's no appropriate way to pass in some parameters for some operations like [How to pass partIndex into QuerySignHTTP for WriteMultipart](https://forum.beyondstorage.io/t/how-to-pass-partindex-into-querysignhttp-for-writemultipart/192).
+- `CreateMultipart` needs to support Expires parameter.
+
+## Proposal
+
+I propose to add `MultipartHTTPSigner` interface.
+
+```go
+type MultipartHTTPSigner interface {
+    QuerySignHTTPCreateMultipart(path string, expire time.Duration, ps ...types.Pairs) (req *http.Request, err error)
+    QuerySignHTTPWriteMultipart(o *Object, size int64, index int, expire time.Duration, ps ...types.Pairs) (req *http.Request, err error)
+    QuerySignHTTPListMultipart(o *Object, expire time.Duration, ps ...types.Pairs) (req *http.Request, err error)
+    QuerySignHTTPCompleteMultipart(o *Object, parts []*Part, expire time.Duration, ps ...types.Pairs) (req *http.Request, err error)
```

Comment:
Where is `o *Object` come from?

Reply:
Oh, but we need the `path` and `uploadId`. How can we pass them in better?

Reply:
I think we need to add return value for `QuerySignHTTPCreateMultipart`, maybe `MultipartID` or `Object`? And also need to add return value for `QuerySignHTTPWriteMultipart`(including `Part`).

Reply:
I think the current design is OK. Users can use `Create` to build an object.

Reply:
`MultipartID` and `Part`(`ETag`) are returned by server. If the current design is kept, we will not be able to use `QuerySignHTTPCreateMultipart` and `QuerySignHTTPWriteMultipart`  at the same time, and the part uploaded by `QuerySignHTTPWriteMultipart` will not be able to merge due to the lack of etag.

Reply:
Oh, I misunderstood it. We just constructed the `*http.Request` for users and the requests are not send so we can't get these infos.

Reply:
Yes, they can't. They need to parse the returning response by themself.

For example:

```go
req, _ := store.QuerySignHTTPCreateMultipart(xxxx)
resp, _ := client.Send(req)
var cmr createMultipartResp
_ := json.Unmarshal(resp.Body, &cmr)

o := store.Create(xxxx, cmr.MustGetMultipartId())
req, _ = store.QuerySignHTTPWriteMultipart(o, xxx)
```